### PR TITLE
[Windows/coverage] Support sndrcvflood on Windows + select.select

### PR DESCRIPTION
**Pull request by gpotter2**
This PR:
- Adds a very basic `select.select` windows alternative for the L2-3Sockets.
- Fix `sndrcvflood` on Windows.
- Adds tests for `sndrcvflood`

### CODE REVIEW

**Comment by gpotter2** [comment]
@guedou Hello, could

**Comment by gpotter2** [comment]
@guedou Hello ! Could you review this please ?

File: scapy/arch/pcapdnet.py [hunk 1]
```diff
@@ -156,9 +156,21 @@ def datalink(self):
           return pcap_datalink(self.pcap)
       def fileno(self):
           if sys.platform.startswith("win"):
-            log_loading.error("Cannot get selectable PCAP fd on Windows")
+            log_loading.error("Cannot get selectable PCAP fd on Windows. "
+                              "Use L2socket/L3socket select() function instead !")
             return 0
-          return pcap_get_selectable_fd(self.pcap) 
+          return pcap_get_selectable_fd(self.pcap)
+      def select(self, socket, timeout):
```
Comment (guedou) [inline scapy/arch/pcapdnet.py]: Could you add a docstring and comments ?

File: scapy/arch/windows/compatibility.py [hunk 2]
```diff
@@ -167,9 +167,64 @@ def sndrcv(pks, pkt, timeout = 2, inter = 0, verbose=None, chainCC=0, retry=0, m
         print "\nReceived %i packets, got %i answers, remaining %i packets" % (nbrecv+len(ans), len(ans), notans)
     return plist.SndRcvList(ans),plist.PacketList(remain,"Unanswered")
 
+def sndrcvflood(pks, pkt, prn=lambda s_r: s_r[1].summary(), chainCC=0, store=1, unique=0, timeout=None):
```
Comment (guedou) [inline scapy/arch/windows/compatibility.py]: Could you add a docstring and comments ?

File: scapy/arch/windows/compatibility.py [hunk 3]
```diff
@@ -167,9 +167,64 @@ def sndrcv(pks, pkt, timeout = 2, inter = 0, verbose=None, chainCC=0, retry=0, m
...
+    if not isinstance(pkt, Gen):
+        pkt = SetGen(pkt)
+    tobesent = [p for p in pkt]
```
Comment (guedou) [inline scapy/arch/windows/compatibility.py]: You can use a generator here
tobesent = (p for p in pkt)

File: scapy/arch/windows/compatibility.py [hunk 4]
```diff
@@ -167,9 +167,64 @@ def sndrcv(pks, pkt, timeout = 2, inter = 0, verbose=None, chainCC=0, retry=0, m
...
+    received = plist.SndRcvList()
+    seen = {}
+
+    hsent={}
```
Comment (guedou) [inline scapy/arch/windows/compatibility.py]: add spaces around `=`

File: scapy/arch/windows/compatibility.py [hunk 5]
```diff
@@ -167,9 +167,64 @@ def sndrcv(pks, pkt, timeout = 2, inter = 0, verbose=None, chainCC=0, retry=0, m
...
+                remain = stoptime-time.time()
+                if remain <= 0:
+                    break
+            pks.send(packets_to_send.next())
+                
+            p = pks.recv(MTU)
+            if p is None:
+                continue
+            h = p.hashret()
+            if h in hsent:
```
Comment (guedou) [inline scapy/arch/windows/compatibility.py]: This is super hard to review. Could you refactor and/or comment ?

File: scapy/arch/pcapdnet.py (hunk 1 above)
Comment (gpotter2) [inline scapy/arch/pcapdnet.py]: This should not be called directly, as located in `_PcapWrapper_pypcap`. The good way would be through `supersocket.py` I'll add comments thoug

File: scapy/arch/windows/compatibility.py (hunk 2 above)
Comment (gpotter2) [inline scapy/arch/windows/compatibility.py]: Again I don't want to add a docstring, as this is used by `srpflood`, `srflood` (there are no docstring too on the linux part.) This function is basically a mix between the windows' `sndrcv `and the linux `sndrcvflood`

File: scapy/arch/windows/compatibility.py (hunk 5 above)
Comment (gpotter2) [inline scapy/arch/windows/compatibility.py]: This part is copy/pasted from https://github.com/secdev/scapy/blob/master/scapy/sendrecv.py#L520

**Comment by p-l-** [comment]
OK, we need to do something about Windows functions. I think it's not a good move to keep different implementations that are just copy & paste of same codes with minor adaptations. This will make the code even harder to maintain. @gpotter2 do you think it would be reasonably achievable?

**Comment by gpotter2** [comment]
Surely... I'll try to isolate a work base or something common between them...

**Comment by p-l-** [comment]
This PR raises one huge question: should we continue to have duplicated code for Windows, or should we, instead, try to factor out the functions like `sr()`?
We already have functionalities that won't work under Windows because of that (`sniff()` with multiple interfaces comes to my mind, there might be more), and we will end up with different behavior or API.

What do you think @guedou @gpotter2?

**Comment by p-l-** [comment]
I'll also look at a Windows/Unix compatible implementation of `sndrcvflood()` after `sniff()`.

**Pull request closed without merging**